### ISSUE TITLE
test=true does not 

### ISSUE THREAD

**Issue by doesitblend** [saltstack/salt#45075]
### Description of Issue/Question
A state using `file.managed` will fail when `test=true` is passed and the directory/file does not exist. This happens even when `makedirs: true` is used. This could be a bug or a feature request.

This brings up the question of desired behavior. Should `test=true` with `makedirs: true` be able to verify whether it can create this directory?

### Setup

Add this state file to your state tree and call it on a Windows host with `test=true` like:

`salt 'windows' state.sls testtruefail test=true`

```
{# Intentionally points to an invalid directory #}
{% set beacon_config = "C:\\salt\\conf\\minion2.d\\beacon.conf" %}

beacon-config-file: 
  file.managed: 
    - name: {{ beacon_config }}
    - source: 'salt://beacons.conf' 
    - template: 'jinja'
    - makedirs: true 
    - listen_in: 
      - cmd: restart_salt_minion

restart_salt_minion:
  cmd.run:
    - name: "C:\\salt\\salt-call.bat --local service.restart salt-minion"
    - shell: powershell
```

### Versions Report

This has been tested on Windows 2008 R2, and Windows 2012 R2.
    Salt Version:
               Salt: 2016.11.8
     
    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 7.1.0
           dateutil: 2.5.3
          docker-py: Not Installed
              gitdb: 0.6.4
          gitpython: 2.0.8
              ioflo: 1.5.5
             Jinja2: 2.9.4
            libgit2: Not Installed
            libnacl: 1.4.5
           M2Crypto: Not Installed
               Mako: 1.0.4
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.17
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:24:40) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.3.8
             PyYAML: 3.11
              PyZMQ: 16.0.1
               RAET: Not Installed
              smmap: 0.9.0
            timelib: 0.2.4
            Tornado: 4.4.1
                ZMQ: 4.1.6
     
    System Versions:
               dist:   
            machine: AMD64
            release: 2008ServerR2
             system: Windows
            version: 2008ServerR2 6.1.7600 SP0 Multiprocessor Free
     

    Salt Version:
               Salt: 2017.7.2
     
    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 10.2.1
           dateutil: 2.6.0
          docker-py: Not Installed
              gitdb: 2.0.3
          gitpython: 2.1.3
              ioflo: Not Installed
             Jinja2: 2.9.6
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.6
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.17
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.4.0
             PyYAML: 3.11
              PyZMQ: 16.0.2
               RAET: Not Installed
              smmap: 2.0.3
            timelib: 0.2.4
            Tornado: 4.5.1
                ZMQ: 4.1.6
     
    System Versions:
               dist:   
             locale: cp1252
            machine: AMD64
            release: 2008ServerR2
             system: Windows
            version: 2008ServerR2 6.1.7600 SP0 Multiprocessor Free
     
    Salt Version:
               Salt: 2017.7.1
     
    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 10.2.1
           dateutil: 2.6.0
          docker-py: Not Installed
              gitdb: 2.0.2
          gitpython: 2.1.3
              ioflo: Not Installed
             Jinja2: 2.9.6
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.6
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.17
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.4.0
             PyYAML: 3.11
              PyZMQ: 16.0.2
               RAET: Not Installed
              smmap: 2.0.3
            timelib: 0.2.4
            Tornado: 4.5.1
                ZMQ: 4.1.6
     
    System Versions:
               dist:   
             locale: cp1252
            machine: AMD64
            release: 2012ServerR2
             system: Windows
            version: 2012ServerR2 6.3.9600  Multiprocessor Free

**Comment by garethgreenaway** [comment]
When `test=True` is passed the `check_managed_changes` is called, which does not appear to take `makedirs` into account.

**Comment by terminalmage** [comment]
I can't reproduce this on 2017.11.8 or the head of 2016.11 and 2017.7 branches. You said it failed, but did not provide the actual failure, which would have been useful to help discover what caused the failure.

@garethgreenaway I don't think that the `makedirs` has anything to do with this, `check_managed_changes()` just checks to see if the destination file already exists, and if it does, checks if running the state would result in changes. It does this by running `get_managed()` on a temp file and comparing it to the existing file. If the file doesn't exist, this function just returns a dictionary saying that the path being managed would be a new file.

**Comment by rallytime** [comment]
@doesitblend Can you follow up here?

**Comment by doesitblend** [comment]
@terminalmage I have not tested this since the release of 2017.7.4. The original error I was receiving is as follows. 

```
win2008-2:
----------
          ID: beacon-config-file
    Function: file.managed
        Name: C:\salt\conf\minion2.d\beacon.conf
      Result: False
     Comment: Unable to manage file: Path not found: C:\salt\conf\minion2.d\beacon.conf
     Started: 16:02:42.315000
    Duration: 63.0 ms
     Changes:   
----------
          ID: restart_salt_minion
    Function: cmd.run
        Name: C:\salt\salt-call.bat --local service.restart salt-minion
      Result: None
     Comment: Command "C:\salt\salt-call.bat --local service.restart salt-minion" would have been executed
     Started: 16:02:43.409000
    Duration: 0.0 ms
     Changes:   

Summary for win2008-2
------------
Succeeded: 1 (unchanged=1)
Failed:    1
------------
Total states run:     2
Total run time:  63.000 ms
ERROR: Minions returned with non-zero exit code
```

**Comment by doesitblend** [comment]
ZD-2066

**Comment by terminalmage** [comment]
OK, can you please test on a newer release?

**Comment by terminalmage** [comment]
Oh, I misread 2017.7.4 as 2017.7.2, .4 is the newest. I'll take another look.

**Comment by terminalmage** [comment]
I still can't reproduce. Maybe this is a Windows-specific issue? I recall at some point there was an issue with multiprocessing in Windows where the `test` value was not being respected, but I thought we worked around that a long time ago.

@twangboy or @dwoz, can you test this? I don't have a Windows box available to me.

**Comment by twangboy** [comment]
I was able to reproduce using 2017.7.2.
Looks like it was fixed in 2017.7.3.

Original issue: https://github.com/saltstack/salt/issues/43328
Fixed by: https://github.com/saltstack/salt/pull/43379

Which went into 2017.7.3.
